### ISSUE TITLE
Bring back SWIG_PYTHON_OUTPUT_TUPLE or equivalent in SWIG 4

### ISSUE THREAD

**Issue by longsleep** [swig/swig#1933]
As part of https://github.com/swig/swig/issues/1340 the `-outputtuple` functionality was removed from SWIG with the only explanation that it is unclear why it was required. 

This functionality is used by Kopano (https://github.com/Kopano-dev/kopano-core/blob/kopanocore-11.0.0/swig/python/Makefile.am#L20) since various functions return multiple list output values which are without this option merged wrongly. 

I noticed that someone else already has complained about the very same thing at https://github.com/casadi/casadi/issues/2629

I am wondering what is the correct solution for this problem with SWIG 4 - for now we are building with SWIG 3, but that cannot continue forever.

**Comment by wsfulton** [comment]
Can you provide a use case for this, a standalone example? As mentioned in 0a9b36d3beb0055ad685583fd8d91c715d8e0d7e it was undocumented and unclear why this is needed. If a reason for this is made clear with a use case, it may be possible to suggest an alternative.

**Comment by ojwb** [comment]
@longsleep Ping?

Without some feedback on this, it's unlikely to get resolved.

**Comment by longsleep** [comment]
@ojwb yeah I realize that. We decided to use swig 3 indefinitely for the time being. I don't know enough about the swig internals to know why it is needed. The kopanocore code contains a pretty large swigged python module which becomes non-functional. Back in the day I tried to get the previous behavior without success. https://github.com/casadi/casadi/issues/2629 has an example of the difference. Maybe someone with more swig experience can make a standalone example which produces multiple list results as return value.

**Comment by ojwb** [comment]
I think the problem is in `SWIG_Python_AppendOutput()`.

The result starts off as `Py_None` or `NULL`.

The first return value is a `PyList`, so now the result set to that `PyList`.

The next return value is a `PyList`.  The return value is already a `PyList` so this `PyList` is appended as a new element to the first `PyList`.

The essential issue is that there's no distinction between a single return value which is a `PyList` and multiple return values in a `PyList`.  To resolve this I think we either need to track how many times `SWIG_Python_AppendOutput` has been called, or generate code which knows how many values are going to be appended.  The latter seems a bit cleaner, but only works if the number of return values is fixed for a given function - if it can vary this approach doesn't work.

@wsfulton Thoughts?

**Comment by ojwb** [comment]
(And the reason this worked with a tuple as the container is presumably just that it's a different type to list.)

**Comment by jgillis** [comment]
@wsfulton -- here is a stub for a unittest that fails since SWIG 4:
```
%module python_argout_list_bug

// Test case for SWIG output parameter bug when OUTPUT typemap returns a list
// Bug: When a void function has multiple OUTPUT parameters that are lists,
// the second list gets appended INTO the first list instead of creating
// a separate tuple/list of results.
//
// Expected: [[0, 2, 4], [1, 3, 5]]
// Buggy:    [0, 2, 4, 1, 3, 5]
//
// This bug occurs because SWIG_Python_AppendOutput() doesn't distinguish
// between an output parameter that IS a list vs needing to create a list
// to hold multiple outputs.

%include "typemaps.i"
%include "std_vector.i"

%template(IntVector) std::vector<int>;

// Define OUTPUT typemap for std::vector<int>&
%typemap(in, numinputs=0) std::vector<int>& OUTPUT (std::vector<int> temp) {
  $1 = &temp;
}

%typemap(argout) std::vector<int>& OUTPUT {
  %append_output(swig::from((*$1)));
}

// Apply OUTPUT typemap to convert reference parameters to return values
%apply std::vector<int>& OUTPUT { std::vector<int>& v1 };
%apply std::vector<int>& OUTPUT { std::vector<int>& v2 };

%inline %{
#include <vector>

// Function that returns two vectors via output parameters
void get_two_vectors(std::vector<int>& v1, std::vector<int>& v2) {
  v1.clear();
  v2.clear();
  v1.push_back(0);
  v1.push_back(2);
  v1.push_back(4);

  v2.push_back(1);
  v2.push_back(3);
  v2.push_back(5);
}

%}
```

I guess this is also fixable without reviving tuple outputs.

**Comment by wsfulton** [comment]
Hi @jgillis. Calling `get_two_vectors()` using your testcase above outputs `[(0, 2, 4), (1, 3, 5)]` consistently for me (tested with swig-3.0.12, swig-4.0.0, swig-4.5.0). So I can't replicate your buggy output `[0, 2, 4, 1, 3, 5]`. 

However, I can replicate a similar sort of problem when `SWIG_PYTHON_OUTPUT_TUPLE` was defined before swig 4 where the output is `[0, 2, 4, [1, 3, 5]]` by modifying your argout typemap to use a List instead of a Tuple as the output from `std::vector`:
```swig
%typemap(argout) std::vector<int>& OUTPUT {
  // Variation that uses lists instead of tuples for vector wrapping
  PyObject *tuple = swig::from((*$1));
  PyObject *list = PySequence_List(tuple);
  Py_DECREF(tuple);
  %append_output(list);
}
```
The removed `SWIG_PYTHON_OUTPUT_TUPLE` definition / `-outputtuple` behaviour can be restored by adding the following at the start of your interface file (before any of other `%include`:
```swig
//Note SWIG's macro definitions...
// #define SWIG_AppendOutput(result, obj)  SWIG_Python_AppendOutput(result, obj, $isvoid)
// #define %append_output(obj)               $result = SWIG_AppendOutput($result,obj) 
//Define a new variant:
#undef SWIG_AppendOutput
#define SWIG_AppendOutput(result, obj)  SWIG_Python_AppendOutput_Tuple(result, obj, $isvoid)

%{
SWIGINTERN PyObject*
SWIG_Python_AppendOutput_Tuple(PyObject* result, PyObject* obj, int is_void) {
  PyObject*   o2;
  PyObject*   o3;
  if (!result) {
    result = obj;
  } else if (result == Py_None && is_void) {
    Py_DECREF(result);
    result = obj;
  } else {
    if (!PyTuple_Check(result)) {
      o2 = result;
      result = PyTuple_New(1);
      PyTuple_SET_ITEM(result, 0, o2);
    }
    o3 = PyTuple_New(1);
    PyTuple_SET_ITEM(o3, 0, obj);
    o2 = result;
    result = PySequence_Concat(o2, o3);
    Py_DECREF(o2);
    Py_DECREF(o3);
  }
  return result;
}
%}
```
Output is then `([0, 2, 4], [1, 3, 5])`.

I've updated `SWIG_Python_AppendOutput_Tuple` from the original removed in 0a9b36d3beb0 to include the fix the improvements added to `SWIG_Python_AppendOutput` described in #2905. But absolutely identical behaviour to swig-3.0.12 and earlier can be obtained by replacing `$isvoid` with 1, as mentioned briefly in the CHANGES file entry dated 2024-10-06.

This is the workaround to use for the few users who are missing the removed option.

**Comment by jgillis** [comment]
fwiw, my test case was indeed lacking. Here a proper version:
```
%module python_argout_list_bug

%include "std_vector.i"

%template(IntVector) std::vector<int>;

%typemap(in, numinputs=0) std::vector<int>& OUTPUT (std::vector<int> temp) {
  $1 = &temp;
}

%typemap(argout) std::vector<int>& OUTPUT {
  PyObject *tuple = swig::from((*$1));
  PyObject *list = PySequence_List(tuple);
  Py_DECREF(tuple);
  %append_output(list);
}

%apply std::vector<int>& OUTPUT { std::vector<int>& v1 };
%apply std::vector<int>& OUTPUT { std::vector<int>& v2 };

%inline %{
#include <vector>

void get_two_lists(std::vector<int>& v1, std::vector<int>& v2) {
  v1.clear();
  v2.clear();
  v1.push_back(0);
  v1.push_back(2);
  v1.push_back(4);

  v2.push_back(1);
  v2.push_back(3);
  v2.push_back(5);
}
%}
```
Your suggestion works perfectly fine; issue can stay closed.
